emcc.py

```diff
@@ -1902,12 +1902,14 @@ def include_and_export(name):
         exit_with_error('MEM_INIT_METHOD is not supported in wasm. Memory will be embedded in the wasm binary if threads are not used, and included in a separate file if threads are used.')
       if shared.Settings.WASM2JS:
         shared.Settings.MAYBE_WASM2JS = 1
-        # wasm2js does not support passive segments or atomics
-        if shared.Settings.USE_PTHREADS:
-          exit_with_error('WASM2JS does not yet support pthreads')
-        # in wasm2js, keep the mem init in the wasm itself if we can and if the
-        # options wouldn't tell a js build to use a separate mem init file
-        shared.Settings.MEM_INIT_IN_WASM = not options.memory_init_file or shared.Settings.SINGLE_FILE
+        # when using wasm2js, if the memory segments are in the wasm then they
+        # end up converted by wasm2js into base64 encoded JS. alternatively, we
+        # can use a .mem file like asm.js used to.
+        # generally we follow what the options tell us to do (which is to use
+        # a .mem file in most cases, since it is binary & compact). however, for
+        # pthreads we must keep the memory segments in the wasm as they will be
+        # passive segments which the .mem format cannot handle.
+        shared.Settings.MEM_INIT_IN_WASM = not options.memory_init_file or shared.Settings.SINGLE_FILE or shared.Settings.USE_PTHREADS
```

Comment:
should we be throwing an error if using wasm2js + pthreads + memory_init_file rather than silently ignoring it? Especially since users and build systems may be expecting a file to be created and may get a stale one?

Reply:
Good question. The downside to erroring here is that it would mean `emcc -pthread -O2` errors, since that defaults to emitting a mem init file, which would be common and more annoying, I think.

We can adjust this later if there are problems in practice. As this option is mainly for debugging, I don't think that would be a problem. (Also, we can adjust it by making a `.mem` file that does support passive segments, if we think it's worth it.)

Another option here is to not emit a mem init file but also ensure none exists, so there is no chance of a stale file. Do you think that's worth it?

Reply:
I think it might be. In general I tend to favor loud failures with easy fixes over making things "just work" if there is a possibility of a subtle failure. I agree the stakes are not super high here though.

Reply:
(on the other hand it would be surprising if the compiler deleted a file, since compilers don't generally do that. But IIUC this is a case where the user is specifying an output filename, and is expecting it to be overwritten, so I think that would be ok.)

Reply:
Hmm, I am more convinced actually that we shouldn't do it - good point that it's odd for a compiler to remove a file.

How about landing this as is, and seeing if it's an issue in practice?

Reply:
We can, I don't feel strongly. But I also think that removing a file isn't that bad when the expectation is that the file will be overwritten.

Reply:
Ok, landing for now as is, since after further thought I think I prefer to go the way that keeps emcc.py code simpler.